### PR TITLE
8303078: Reduce allocations when pretty printing JCTree during compilation

### PR DESCRIPTION
**Pull request by dreis2211**
Hi,

I've been recently optimizing a few project pipelines and always noticed that their compilation step is somewhat "expensive".  Zooming into the issue always revealed Lombok to be a larger contributor. Unfortunately, I can't get rid of Lombok in this customer project (before you ask).

Apparently, Lombok is printing the respective `JCTree` here and there to check for type matches. I can't imagine this to be super efficient, but that's how it is at the moment.
image

Anyhow, regardless of the Lombok inefficiencies I think there are some optimization opportunities in the JDK itself. 

1. Overall, `Pretty.visitSelect` accounts for 8-10% of the total allocations in this project. And among those there are StringBuilder allocations coming from the following:

```
    public void visitSelect(JCFieldAccess tree) {
        try {
            printExpr(tree.selected, TreeInfo.postfixPrec);
            // StringBuilder allocations hiding in here.
            print("." + tree.name);
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }
```

This PR splits the `print` calls into two separate ones to avoid this String concatenation.
```
...
            printExpr(tree.selected, TreeInfo.postfixPrec);
            print('.');
            print(tree.name);
...
```

Secondly, the `print` method takes an `Object` which seems like a good fit for another (private?) variant of it that only takes a `char`. By this we would probably avoid any eventual boxing and avoid any conversion with `Convert.escapeUnicode(s.toString())` that seems superfluous for chars like `.`, ` `, or any braces like `(`, `{` etc. 

This is currently a draft PR as long as the scope is not clarified. It currently only includes the necessary changes that would optimize the particular use-case. But there are more cases where e.g. the new `char` variant could be used and/or any String concatenation could be split into separate `print` calls.

Let me know what you think and if I should include the other cases as well. If you think this is worthwhile, I'd appreciate if this is is sponsored. (Including creating an issue as I can't do this myself apparently. I will of course squash everything together with the proper issue ID once available.)

I've contributed before, so the CLA should be signed.

Cheers,
Christoph

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8303078](https://bugs.openjdk.org/browse/JDK-8303078): Reduce allocations when pretty printing JCTree during compilation

### Reviewers
 * [Jonathan Gibbons](https://openjdk.org/census#jjg) (@jonathan-gibbons - **Reviewer**)
 * [Vicente Romero](https://openjdk.org/census#vromero) (@vicente-romero-oracle - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12667/head:pull/12667` \
`$ git checkout pull/12667`

Update a local copy of the PR: \
`$ git checkout pull/12667` \
`$ git pull https://git.openjdk.org/jdk pull/12667/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12667`

View PR using the GUI difftool: \
`$ git pr show -t 12667`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12667.diff">https://git.openjdk.org/jdk/pull/12667.diff</a>

</details>

### CODE REVIEW

**Comment by dreis2211** [comment]
With the patch applied I can reduce things for an internal product.

**Some specs about the project:**

~2500 classes
Average size: ~76 lines
Largest class: ~2500 lines
Mostly hand-written, but some auto-generated protobuf files (I'd say 90/10 split)

**CPU profile before**
image

**CPU profile after**
image

**Allocation profile before**
image

**Allocation profile after**
image

| Mode | Frames matched before | Frames matched after |
|--- | --- | --- |
|`cpu` | 2.27% | 1.42% |
| `alloc` | 7.09% | 4.13% |

In terms of timings I couldn't measure anything substantially on the actual project. Neither a regression, nor a substantial improvement. But it's not so much about time anyhow but about the allocations here. The overall goal is to reduce impact on GC that is using 27% of the CPU frames (to give you some more background). Obviously, this is not coming from the JDK alone and Lombok is a major contributor here, but the JDK can easy the impact of Lombok here I'd say.
image

Admittedly, it's more on the smaller end of things and a smaller knob in terms of GC improvements.

Let me know if you need more information.

**Comment by dreis2211** [comment]
In terms of some more absolute allocation numbers:

**Before (TOP 3 => ~5,6GB)**
```
       bytes  percent  samples  top
  ----------  -------  -------  ---
  3581028736   26.79%    10682  byte[]
  1291159216    9.66%     3930  java.lang.Object[]
   740065552    5.54%     2161  java.lang.String
```
**After (TOP 3 => ~5,4GB)**
```
      bytes  percent  samples  top
  ----------  -------  -------  ---
  3376915824   26.33%    11615  byte[]
  1284160152   10.01%     3847  java.lang.Object[]
   730100712    5.69%     2023  java.lang.String
```

`StringBuilder` savings account for an additional 62MB

File: src/jdk.compiler/share/classes/com/sun/tools/javac/tree/Pretty.java
Comment (jonathan-gibbons) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/tree/Pretty.java:139]: This bypasses the `Convert.escapeUnicode` code on line 132.   That's not bad, but should be clearly documented ... in that this is a special method for known ASCII characters, and not a general purpose method that could be used on the characters of a string

**Review by jonathan-gibbons — COMMENTED** [review]
If we proceed with this, I would recommend looking for other possible uses of the new method.

A simple `grep` shows 129 single character strings, of which 97 are arguments to `print`.

Separate but related, I note that a similar class, `DocPretty.java`, has 75 single character strings, of which  68 are arguments to print.

Overall changing the coding style to use `out.print(char)` seems worthwhile.

**Comment by dreis2211** [comment]
@jonathan-gibbons Thanks for your review.
> If we proceed with this, I would recommend looking for other possible uses of the new method.
That's the idea - there are multiple cases where this could be used. I'll push an updated version that already changes things to make reviewing a bit easier.

File: src/jdk.compiler/share/classes/com/sun/tools/javac/tree/Pretty.java
Comment (dreis2211) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/tree/Pretty.java:139]: Adjusted the comment in that regards.

**Comment by jonathan-gibbons** [comment]
Filed [JDK-8303078](https://bugs.openjdk.org/browse/JDK-8303078)

**Comment by dreis2211** [comment]
@jonathan-gibbons Made the adjustments to `Pretty`.

**Review by vicente-romero-oracle — COMMENTED** [review]
this is looking good, I think that once the changes to `DocPretty.java`, as Jon suggested, are included in this patch it should be good to go

**Comment by dreis2211** [comment]
@vicente-romero-oracle & @jonathan-gibbons made the adjustments to `DocPretty` now as well.

File: src/jdk.compiler/share/classes/com/sun/tools/javac/tree/Pretty.java
Comment (vicente-romero-oracle) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/tree/Pretty.java:1246]: nit, just below: `String sep = "";` -> `String sep = '\0';` also `sep = ","` -> sep = ','?

File: src/jdk.compiler/share/classes/com/sun/tools/javac/tree/Pretty.java
Comment (vicente-romero-oracle) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/tree/Pretty.java:1456]: `print('\'');` ?

File: src/jdk.compiler/share/classes/com/sun/tools/javac/tree/Pretty.java
Comment (dreis2211) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/tree/Pretty.java:1246]: Done

File: src/jdk.compiler/share/classes/com/sun/tools/javac/tree/Pretty.java
Comment (dreis2211) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/tree/Pretty.java]: Done

File: src/jdk.compiler/share/classes/com/sun/tools/javac/tree/Pretty.java
Comment (dreis2211) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/tree/Pretty.java]: Already resolved

File: src/jdk.compiler/share/classes/com/sun/tools/javac/tree/DocPretty.java
Comment (dreis2211) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/tree/DocPretty.java:195]: Done - but had to change the one with the `sep` argument to take a char instead of String. Seems not used elsewhere, but wasn't sure because of the `protected`

File: src/jdk.compiler/share/classes/com/sun/tools/javac/tree/Pretty.java
Comment (dreis2211) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/tree/Pretty.java]: Done

File: src/jdk.compiler/share/classes/com/sun/tools/javac/tree/Pretty.java
Comment (dreis2211) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/tree/Pretty.java]: Done

File: src/jdk.compiler/share/classes/com/sun/tools/javac/tree/DocPretty.java
Comment (jonathan-gibbons) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/tree/DocPretty.java:195]: While `DocPretty` is public, the package is not publicly exported from the `jdk.compiler` module, and there are no subtypes within the JDK code base, so I think we can ignore the `protected` for now.

If we wanted to be paranoid, it would be reasonable to change `print(char)` to something along the lines of
```
if (Character.isASCII(ch)) out.write(ch); else print((Character) ch);
```

Comment (dreis2211) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/tree/DocPretty.java:195]: Should we (I in that case) be paranoid, though? I would leave it as-is. There is a big class comment above that anyone depending on the internal api uses this at his own risk.

File: src/jdk.compiler/share/classes/com/sun/tools/javac/tree/DocPretty.java
Comment (jonathan-gibbons) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/tree/DocPretty.java:82]: The restriction to skip calling `Convert.escapeUnicode(s.toString())` leaks into `print(List, char sep)`.

If we wanted to be paranoid we could do one or more of the following;

* Change the body of this method to something like<br> `if (Character.isASCII(c)) out.write(c); else print((Character) c);`
* include an `assert` or call to a method in `Assert.*` to verify the argument is ASCII.
* change that `protected void print(List, char)` method to be `private` instead of `protected`

Comment (dreis2211) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/tree/DocPretty.java:82]: Already asked above: should we really be paranoid here? There is a big class comment above that anyone depending on the internal api uses this at his own risk...
But if I should change anything I'd rather make the method `private`

File: src/jdk.compiler/share/classes/com/sun/tools/javac/tree/DocPretty.java
Comment (jonathan-gibbons) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/tree/DocPretty.java:97]: See comment above, suggesting that we reduce access to `private`.

Comment (dreis2211) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/tree/DocPretty.java]: Done

File: src/jdk.compiler/share/classes/com/sun/tools/javac/tree/DocPretty.java
Comment (jonathan-gibbons) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/tree/DocPretty.java:82]: _(Sorry for the duplicate entry)_

I think making the method `private` is good enough, and doing it in the same changeset that changes the signature is even better for future code historians.

**Comment by dreis2211** [comment]
There are some test failures in langtools/tier1 that seem related. Will check

**Comment by jonathan-gibbons** [comment]
I confirm that all the langtools tests now pass, at least on my local laptop.

**Review by jonathan-gibbons — APPROVED** [review]
Thanks for updating `DocPretty.java` to follow the same conventions as `Pretty.java`, even though it was probably not required for your Lombok-based use-case.

I've gone through all the latest patch, and run all the langtools tests as well. All pass on my laptop, and there is no reason to expect any cross-platform issues.

I'll mark it Approved, but while not strictly necessary, I recommend that you wait for Vicente's approval as well, in case he has any final comments.

**Comment by vicente-romero-oracle** [comment]
nit please consider including this minor change:

```
diff --git a/src/jdk.compiler/share/classes/com/sun/tools/javac/tree/Pretty.java b/src/jdk.compiler/share/classes/com/sun/tools/javac/tree/Pretty.java
index f87d5c05a74..dd05d4ac9d1 100644
--- a/src/jdk.compiler/share/classes/com/sun/tools/javac/tree/Pretty.java
+++ b/src/jdk.compiler/share/classes/com/sun/tools/javac/tree/Pretty.java
@@ -1244,11 +1244,11 @@ public class Pretty extends JCTree.Visitor {
             if (tree.paramKind == JCLambda.ParameterKind.EXPLICIT) {
                 printExprs(tree.params);
             } else {
-                String sep = "";
+                char sep = '\0';
                 for (JCVariableDecl param : tree.params) {
                     print(sep);
                     print(param.name);
-                    sep = ",";
+                    sep = ',';
                 }
             }
             print(")->");
```

thanks!

**Review by vicente-romero-oracle — APPROVED** [review]
looks good. I will run an internal test that executes all regression tests in all supported platforms just to be xtra sure

**Comment by vicente-romero-oracle** [comment]
> looks good. I will run an internal test that executes all regression tests in all supported platforms just to be xtra sure

all green :)

File: src/jdk.compiler/share/classes/com/sun/tools/javac/tree/Pretty.java
Comment (dreis2211) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/tree/Pretty.java:1246]: This failed the build - I reverted it.

**Comment by dreis2211** [comment]
Thanks @jonathan-gibbons & @vicente-romero-oracle for your active advice, reviewing and sponsoring :)

**Comment by jonathan-gibbons** [comment]
> Thanks @jonathan-gibbons & @vicente-romero-oracle for your active advice, reviewing and sponsoring :)

You're welcome; thanks for the contribution.

**Comment by vicente-romero-oracle** [comment]
> > Thanks @jonathan-gibbons & @vicente-romero-oracle for your active advice, reviewing and sponsoring :)
> 
> You're welcome; thanks for the contribution.

+1

**Pull request closed without merging**
